### ISSUE TITLE
Provide a way to catch that cleartext HTTP traffic was attempted

### ISSUE THREAD

**Issue by klyubin** [square/okhttp#1392]
This is a follow-up to feature request 1058 where okhttp offers a way to disallow cleartext traffic. Just blocking cleartext traffic is insufficient for some apps. Some app developers want to not only block cleartext traffic but also report to their servers that their app attempted cleartext traffic. This is so that regressions from HTTPS to cleartext HTTP can be investigated and fixed quicker.

At the moment, when okhttp blocks a cleartext HTTP request, it throws a new SocketException("No route to ... ; exhausted connection specs: ..."). Please provide a way to differentiate the exception thrown in this case from other IOExceptions. For example, when blocking cleartext HTTP okhttp could throw its own special subclass of IOException or perhaps a java.lang.SecurityException or an IOException caused by a SecurityException.

I'm happy to implement the changes once a decision reached about what exception to throw.

**Comment by swankjesse** [comment]
`UnknownServiceException` ?

**Comment by swankjesse** [comment]
```
public class UnknownServiceException
extends IOException
```

> Thrown to indicate that an unknown service exception has occurred. Either the MIME type returned by a URL connection does not make sense, or the application is attempting to write to a read-only URL connection.

**Comment by klyubin** [comment]
UnknownServiceException's name implies it's about the the user/client of the API requesting something that's not known by the implementor. However, the Javadoc of this exception says it can also be thrown if you attempt to write over a read-only connection. It's weird and doesn't sound like a great fit. Then again, we might not have too much choice.

Why not SecurityException which is about violating access/security policy? Once argument against that I've heard so far is that it's not a checked type and thus it might crash the process because it's not being caught (usually). However, code that cares about the security implications can catch these exceptions. Also, SecurityException is what happens if you have a Security Manager installed that denies TCP connections to some destination. Thus, if cleartext HTTP is denied, it's somewhat natural that SecurityException is thrown.

We could also let the user of okhttp customize what exception is thrown. For example, they could provide the class of the exception. This way, they can choose themselves whether they want a SecurityException or some subclass of an IOException.

**Comment by klyubin** [comment]
P.S. As I see it, the main difference from the API user's perspective between SecurityException vs IOException is that the former crashes the process by default whereas the latter does not. So, if you're only interested in blocking cleartext HTTP traffic but not interested in knowing that it was blocked and your code can continue working if cleartext HTTP requests fail, you might prefer an IOException or similar.

**Comment by swankjesse** [comment]
Let's try to do this with minimal new API, 'cause I don't think most users are going to benefit from this. How about `UnknownServiceException` ? It's not a perfect fit, but it's in the spirit of what we're trying to do. And it means we don't need to introduce a new exception type in our API.

**Comment by klyubin** [comment]
UnknownServiceException it is then. What about the API for enabling this behavior? In the discussion of 1058 we talked about adding a method that states whether cleartext should be permitted. But later this was changed into specifying the set of acceptable routes. Shall I just modify the code such that when cleartext HTTP is requested by there's no CLEARTEXT route in the set, then the UnknownServiceException is thrown? This would remove the need to expose any additional API.
